docs/man/btlaunchmanycurses.bittornado.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.33.
.TH BTLAUNCHMANYCURSES "1" "August 2004" "btlaunchmanycurses (bittornado)"
.SH NAME
Btlaunchmanycurses \- launch multi torrent downloads, curses interface
.SH SYNOPSIS
.B btlaunchmanycurses
\fI<directory> <global options>\fR
.SH DESCRIPTION
Launches a separate torrent download thread for each .torrent file in
the directory specified by <directory>, using the curses interface.
.SH OPTIONS
This program follows the usual GNU command-line syntax, with long options
starting with two dashes ('-'). A summary of options is included below.
.PP
<directory> - directory to look for .torrent files (semi-recursive)
.PP
\fB\-\-max_uploads\fR \fInumber\fP
.IP
the maximum \fInumber\fP of uploads to allow at once. (defaults to 7)
.PP
\fB\-\-keepalive_interval\fR \fIseconds\fP
.IP
number of \fIseconds\fP to pause between sending keepalives (defaults to 120.0)
.PP
\fB\-\-download_slice_size\fR \fIbytes\fP
.IP
How many \fIbytes\fP to query for per request. (defaults to 16384)
.PP
\fB\-\-upload_unit_size\fR \fIbytes\fP
.IP
when limiting upload rate, how many \fIbytes\fP to send at a time (defaults to 1460)
.PP
\fB\-\-request_backlog\fR \fInumber\fP
.IP
maximum \fInumber\fP of requests to keep in a single pipe at once. (defaults to 10)
.PP
\fB\-\-max_message_length\fR \fIlength\fP
.IP
maximum \fIlength\fP prefix encoding you'll accept over the wire - larger values get the
connection dropped. (defaults to 8388608)
.PP
\fB\-\-ip\fR \fIip\fP
.IP
\fIip\fP to report you have to the tracker. (defaults to '')
.PP
\fB\-\-minport\fR \fIportnum\fP
.IP
set \fIportnum\fP as the minimum port to listen on, counts up if unavailable (defaults to 10000)
.PP
\fB\-\-maxport\fR \fIportnum\fP
.IP
set \fIportnum\fP as the maximum port to listen on (defaults to 60000)
.PP
\fB\-\-random_port\fR \fI 0 | 1 \fP
.IP
whether to choose randomly inside the port range instead of counting up linearly
(defaults to 1)
.PP
\fB\-\-responsefile\fR \fIfile\fP
.IP
\fIfile\fP the server response was stored in, alternative to url (defaults to '')
.PP
\fB\-\-url\fR \fIURL\fP
.IP
\fIURL\fP to get file from, alternative to responsefile (defaults to '')
.PP
\fB\-\-selector_enabled\fR \fI 0 | 1 \fP
.IP
whether to enable the file selector and fast resume function (defaults to 1)
.PP
\fB\-\-expire_cache_data\fR \fIdays\fP
.IP
the number of \fIdays\fP after which you wish to expire old cache data (0 = disabled) (defaults
to 10)
.PP
\fB\-\-priority\fR \fI -1|0|1|2[,-1|0|1|2] \fP
.IP
a list of file priorities separated by commas, must be one per file, 0 = highest, 1 =
normal, 2 = lowest, \fB\-1\fR = download disabled (defaults to ''). Order is based
on the file/torrent order as shown by btshowmetainfo. For example, to download only
the third of four files use: --priority -1,-1,2,-1
.PP
\fB\-\-saveas\fR \fIfilename\fP
.IP
local \fIfilename\fP to save the file as, null indicates query user (defaults to '')
.PP
\fB\-\-timeout\fR \fIseconds\fP
.IP
time in \fIseconds\fP to wait between closing sockets which nothing has been received on (defaults to
300.0)
.PP
\fB\-\-timeout_check_interval\fR \fIseconds\fP
.IP
time to wait in \fIseconds\fP between checking if any connections have timed out (defaults to 60.0)
.PP
\fB\-\-max_slice_length\fR \fIlength\fP
.IP
maximum \fIlength\fP slice to send to peers, larger requests are ignored (defaults to 131072)
.PP
\fB\-\-max_rate_period\fR \fIseconds\fP
.IP
maximum amount of time in \fIseconds\fP to guess the current rate estimate represents (defaults to 20.0)
.PP
\fB\-\-bind\fR \fIip[,hostname]\fP
.IP
comma-separated list of \fIips/hostnames\fP to bind to locally (defaults to '')
.PP
\fB\-\-ipv6_enabled\fR \fI 0 | 1 \fP
.IP
allow the client to connect to peers via IPv6 (defaults to 0)
.PP
\fB\-\-ipv6_binds_v4\fR \fI 0 | 1 \fP
.IP
set if an IPv6 server socket will also field IPv4 connections (defaults to 1)
.PP
\fB\-\-upnp_nat_access\fR \fI 0 | 1 | 2 \fP
.IP
attempt to autoconfigure a UPnP router to forward a server port (0 = disabled, 1 = mode 1
[fast], 2 = mode 2 [slow]) (defaults to 1)
.PP
\fB\-\-upload_rate_fudge\fR \fIseconds\fP
.IP
time equivalent in \fIseconds\fP of writing to kernel-level TCP buffer, for rate adjustment (defaults to
5.0)
.PP
\fB\-\-tcp_ack_fudge\fR \fIoverhead\fP
.IP
how much TCP ACK download \fIoverhead\fP to add to upload rate calculations (0 = disabled)
(defaults to 0.029999999999999999)
.PP
\fB\-\-display_interval\fR \fIseconds\fP
.IP
time in \fIseconds\fP between updates of displayed information (defaults to 0.5)
.PP
\fB\-\-rerequest_interval\fR \fIseconds\fP
.IP
time to wait, in \fIseconds\fP, between requesting more peers (defaults to 300)
.PP
\fB\-\-min_peers\fR \fInumber\fP
.IP
minimum \fInumber\fP of peers to not do rerequesting (defaults to 20)
.PP
\fB\-\-http_timeout\fR \fIseconds\fP
.IP
number of \fIseconds\fP to wait before assuming that an http connection has timed out (defaults
to 60)
.PP
\fB\-\-max_initiate\fR \fInumber\fP
.IP
\fInumber\fP of peers at which to stop initiating new connections (defaults to 40)
.PP
\fB\-\-check_hashes\fR \fI 0 | 1 \fP
.IP
whether to check hashes on disk (defaults to 1)
.PP
\fB\-\-max_upload_rate\fR \fIkB/s\fP
.IP
maximum \fIkB/s\fP to upload at (0 = no limit, \fB\-1\fR = automatic) (defaults to 0)
.PP
\fB\-\-max_download_rate\fR \fIkB/s\fP
.IP
maximum \fIkB/s\fP to download at (0 = no limit) (defaults to 0)
.PP
\fB\-\-alloc_type\fR \fInormal | background | pre-allocate | sparse\fP
.IP
allocation type (may be normal, background, pre-allocate or sparse) (defaults to
\&'normal')
.PP
\fB\-\-alloc_rate\fR \fIMiB/s\fP
.IP
rate (in \fIMiB/s\fP) to allocate space at using background allocation (defaults to 2.0)
.PP
\fB\-\-buffer_reads\fR \fI 0 | 1 \fP
.IP
whether to buffer disk reads (defaults to 1)
.PP
\fB\-\-write_buffer_size\fR \fIspace\fP
.IP
the maximum amount of \fIspace\fP to use for buffering disk writes (in megabytes, 0 = disabled)
(defaults to 4)
.PP
\fB\-\-snub_time\fR \fIseconds\fP
.IP
\fIseconds\fP to wait for data to come in over a connection before assuming it's
semi-permanently choked (defaults to 30.0)
.PP
\fB\-\-spew\fR \fI 0 | 1 \fP
.IP
whether to display diagnostic info to stdout (defaults to 0)
.PP
\fB\-\-rarest_first_cutoff\fR \fInumber\fP
.IP
\fInumber\fP of downloads at which to switch from random to rarest first (defaults to 2)
.PP
\fB\-\-rarest_first_priority_cutoff\fR \fInumber\fP
.IP
the \fInumber\fP of peers which need to have a piece before other partials take priority over
rarest first (defaults to 5)
.PP
\fB\-\-min_uploads\fR \fInumber\fP
.IP
the \fInumber\fP of uploads to fill out to with extra optimistic unchokes (defaults to 4)
.PP
\fB\-\-max_files_open\fR \fInumber\fP
.IP
the maximum \fInumber\fP of files to keep open at a time, 0 means no limit (defaults to 50)
.PP
\fB\-\-round_robin_period\fR \fIseconds\fP
.IP
the number of \fIseconds\fP between the client's switching upload targets (defaults to 30)
.PP
\fB\-\-super_seeder\fR \fI 0 | 1 \fP
.IP
whether to use special upload-efficiency-maximizing routines (only for dedicated seeds)
(defaults to 0)
.PP
\fB\-\-security\fR \fI 0 | 1 \fP
.IP
whether to enable extra security features intended to prevent abuse (defaults to 1)
.PP
\fB\-\-max_connections\fR \fInumber\fP
.IP
the absolute maximum \fInumber\fP of peers to connect with (0 = no limit) (defaults to 0)
.PP
\fB\-\-auto_kick\fR \fI 0 | 1 \fP
.IP
whether to allow the client to automatically kick/ban peers that send bad data (defaults
to 1)
.PP
\fB\-\-double_check\fR \fI 0 | 1 \fP
.IP
whether to double-check data being written to the disk for errors (may increase CPU load)
(defaults to 1)
.PP
\fB\-\-triple_check\fR \fI 0 | 1 \fP
.IP
whether to thoroughly check data being written to the disk (may slow disk access)
(defaults to 0)
.PP
\fB\-\-lock_files\fR \fI 0 | 1 \fP
.IP
whether to lock files the client is working with (defaults to 1)
.PP
\fB\-\-lock_while_reading\fR \fI 0 | 1 \fP
.IP
whether to lock access to files being read (defaults to 0)
.PP
\fB\-\-auto_flush\fR \fIminutes\fP
.IP
\fIminutes\fP between automatic flushes to disk (0 = disabled) (defaults to 0)
.PP
\fB\-\-parse_dir_interval\fR \fIseconds\fP
.IP
how often to rescan the torrent directory, in \fIseconds\fP (defaults to
60)
.PP
\fB\-\-saveas_style\fR \fI 1 | 2 | 3 \fP
.IP
How to name torrent downloads (1 = rename to torrent name, 2 = save
under name in torrent, 3 = save in directory under torrent name)
(defaults to 2)
.PP
\fB\-\-display_path\fR \fI 0 | 1 \fP
.IP
whether to display the full path or the torrent contents for each
torrent (defaults to 0)
.PP
\fB\-\-save_options\fR \fI 0 | 1 \fP
.IP
whether to save the current options as the new default configuration
(only for btlaunchmanycurses.py) (defaults to 0)
.PP
.SH SEE ALSO
.BR bittorrent-downloader (1),
.BR btmakemetafile (1).
.br
.SH AUTHOR
This manual page was written by Micah Anderson <micah@riseup.net>,
for the Debian GNU/Linux system (but may be used by others).



btlaunchmanycurses.py
#!/usr/bin/env python

# Written by John Hoffman
# see LICENSE.txt for license information

DOWNLOAD_SCROLL_RATE = 1

from BitTornado import PSYCO
if PSYCO.psyco:
    try:
        import psyco
        assert psyco.__version__ >= 0x010100f0
        psyco.full()
    except:
        pass

from BitTornado.launchmanycore import LaunchMany
from BitTornado.download_bt1 import defaults, get_usage
from BitTornado.parseargs import parseargs
from threading import Event
from sys import argv, exit
from time import time, localtime, strftime
import sys, os
from BitTornado import version, report_email
from BitTornado.ConfigDir import ConfigDir

try:
    import curses
    import curses.panel
    from curses.wrapper import wrapper as curses_wrapper
    from signal import signal, SIGWINCH 
except:
    print 'Textmode GUI initialization failed, cannot proceed.'
    print
    print 'This download interface requires the standard Python module ' \
       '"curses", which is unfortunately not available for the native ' \
       'Windows port of Python. It is however available for the Cygwin ' \
       'port of Python, running on all Win32 systems (www.cygwin.com).'
    print
    print 'You may still use "btdownloadheadless.py" to download.'
    sys.exit(1)

assert sys.version >= '2', "Install Python 2.0 or greater"
try:
    True
except:
    True = 1
    False = 0

Exceptions = []


def fmttime(n):
    if n <= 0:
        return None
    try:
        n = int(n)
        assert n < 5184000  # 60 days
    except:
        return 'connecting to peers'
    m, s = divmod(n, 60)
    h, m = divmod(m, 60)
    return 'ETA in %d:%02d:%02d' % (h, m, s)

def fmtsize(n):
    n = long(n)
    unit = [' B', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB', 'ZiB', 'YiB']
    i = 0
    if (n > 999):
        i = 1
        while i + 1 < len(unit) and (n >> 10) >= 999:
            i += 1
            n >>= 10
        n = float(n) / (1 << 10)
    if i > 0:
        size = '%.1f' % n + '%s' % unit[i]
    else:
        size = '%.0f' % n + '%s' % unit[i]
    return size

def ljust(s, size):
    s = s[:size]
    return s + (' '*(size-len(s)))

def rjust(s, size):
    s = s[:size]
    return (' '*(size-len(s)))+s


class CursesDisplayer:
    def __init__(self, scrwin):
        self.messages = []
        self.scroll_pos = 0
        self.scroll_time = 0
        
        self.scrwin = scrwin
        signal(SIGWINCH, self.winch_handler)
        self.changeflag = Event()
        self._remake_window()

    def winch_handler(self, signum, stackframe):
        self.changeflag.set()
        curses.endwin()
        self.scrwin.refresh()
        self.scrwin = curses.newwin(0, 0, 0, 0)
        self._remake_window()
        self._display_messages()

    def _remake_window(self):
        self.scrh, self.scrw = self.scrwin.getmaxyx()
        self.scrpan = curses.panel.new_panel(self.scrwin)
        self.mainwinh = int(2*(self.scrh)/3)
        self.mainwinw = self.scrw - 4  # - 2 (bars) - 2 (spaces)
        self.mainwiny = 2         # + 1 (bar) + 1 (titles)
        self.mainwinx = 2         # + 1 (bar) + 1 (space)
        # + 1 to all windows so we can write at mainwinw

        self.mainwin = curses.newwin(self.mainwinh, self.mainwinw+1,
                                     self.mainwiny, self.mainwinx)
        self.mainpan = curses.panel.new_panel(self.mainwin)
        self.mainwin.scrollok(0)
        self.mainwin.nodelay(1)

        self.headerwin = curses.newwin(1, self.mainwinw+1,
                                       1, self.mainwinx)
        self.headerpan = curses.panel.new_panel(self.headerwin)
        self.headerwin.scrollok(0)

        self.totalwin = curses.newwin(1, self.mainwinw+1,
                                      self.mainwinh+1, self.mainwinx)
        self.totalpan = curses.panel.new_panel(self.totalwin)
        self.totalwin.scrollok(0)

        self.statuswinh = self.scrh-4-self.mainwinh
        self.statuswin = curses.newwin(self.statuswinh, self.mainwinw+1,
                                       self.mainwinh+3, self.mainwinx)
        self.statuspan = curses.panel.new_panel(self.statuswin)
        self.statuswin.scrollok(0)

        try:
            self.scrwin.border(ord('|'),ord('|'),ord('-'),ord('-'),ord(' '),ord(' '),ord(' '),ord(' '))
        except:
            pass
        self.headerwin.addnstr(0, 2, '#', self.mainwinw - 25, curses.A_BOLD)
        self.headerwin.addnstr(0, 4, 'Filename', self.mainwinw - 25, curses.A_BOLD)
        self.headerwin.addnstr(0, self.mainwinw - 24, 'Size', 4, curses.A_BOLD)
        self.headerwin.addnstr(0, self.mainwinw - 18, 'Download', 8, curses.A_BOLD)
        self.headerwin.addnstr(0, self.mainwinw -  6, 'Upload', 6, curses.A_BOLD)
        self.totalwin.addnstr(0, self.mainwinw - 27, 'Totals:', 7, curses.A_BOLD)

        self._display_messages()
        
        curses.panel.update_panels()
        curses.doupdate()
        self.changeflag.clear()


    def _display_line(self, s, bold = False):
        if self.disp_end:
            return True
        line = self.disp_line
        self.disp_line += 1
        if line < 0:
            return False
        if bold:
            self.mainwin.addnstr(line, 0, s, self.mainwinw, curses.A_BOLD)
        else:
            self.mainwin.addnstr(line, 0, s, self.mainwinw)
        if self.disp_line >= self.mainwinh:
            self.disp_end = True
        return self.disp_end

    def _display_data(self, data):    
        if 3*len(data) <= self.mainwinh:
            self.scroll_pos = 0
            self.scrolling = False
        elif self.scroll_time + DOWNLOAD_SCROLL_RATE < time():
            self.scroll_time = time()
            self.scroll_pos += 1
            self.scrolling = True
            if self.scroll_pos >= 3*len(data)+2:
                self.scroll_pos = 0

        i = int(self.scroll_pos/3)
        self.disp_line = (3*i)-self.scroll_pos
        self.disp_end = False

        while not self.disp_end:
            ii = i % len(data)
            if i and not ii:
                if not self.scrolling:
                    break
                self._display_line('')
                if self._display_line(''):
                    break
            ( name, status, progress, peers, seeds, seedsmsg, dist,
              uprate, dnrate, upamt, dnamt, size, t, msg ) = data[ii]
            t = fmttime(t)
            if t:
                status = t
            name = ljust(name,self.mainwinw-32)
            size = rjust(fmtsize(size),8)
            uprate = rjust('%s/s' % fmtsize(uprate),10)
            dnrate = rjust('%s/s' % fmtsize(dnrate),10)
            line = "%3d %s%s%s%s" % (ii+1, name, size, dnrate, uprate)
            self._display_line(line, True)
            if peers + seeds:
                datastr = '    (%s) %s - %s up %s dn - %s peers %s seeds %.3f dist copies' % (
                                progress, status,
                                fmtsize(upamt), fmtsize(dnamt),
                                peers, seeds, dist )
            else:
                datastr = '    (%s) %s - %s up %s dn' % (
                                progress, status,
                                fmtsize(upamt), fmtsize(dnamt) )
            self._display_line(datastr)
            self._display_line('    '+ljust(msg,self.mainwinw-4))
            i += 1

    def display(self, data):
        if self.changeflag.isSet():
            return

        inchar = self.mainwin.getch()
        if inchar == 12: # ^L
            self._remake_window()

        self.mainwin.erase()
        if data:
            self._display_data(data)
        else:
            self.mainwin.addnstr( 1, int(self.mainwinw/2)-5,
                                  'no torrents', 12, curses.A_BOLD )
        totalup = 0
        totaldn = 0
        for ( name, status, progress, peers, seeds, seedsmsg, dist,
              uprate, dnrate, upamt, dnamt, size, t, msg ) in data:
            totalup += uprate
            totaldn += dnrate
        
        totalup = '%s/s' % fmtsize(totalup)
        totaldn = '%s/s' % fmtsize(totaldn)
        
        self.totalwin.erase()
        self.totalwin.addnstr(0, self.mainwinw-27, 'Totals:', 7, curses.A_BOLD)
        self.totalwin.addnstr(0, self.mainwinw-20 + (10-len(totaldn)),
                              totaldn, 10, curses.A_BOLD)
        self.totalwin.addnstr(0, self.mainwinw-10 + (10-len(totalup)),
                              totalup, 10, curses.A_BOLD)

        curses.panel.update_panels()
        curses.doupdate()

        return inchar in (ord('q'),ord('Q'))

    def message(self, s):
        self.messages.append(strftime('%x %X - ',localtime(time()))+s)
        self._display_messages()

    def _display_messages(self):
        self.statuswin.erase()
        winpos = 0
        for s in self.messages[-self.statuswinh:]:
            self.statuswin.addnstr(winpos, 0, s, self.mainwinw)
            winpos += 1
        curses.panel.update_panels()
        curses.doupdate()

    def exception(self, s):
        Exceptions.append(s)
        self.message('SYSTEM ERROR - EXCEPTION GENERATED')



def LaunchManyWrapper(scrwin, config):
    LaunchMany(config, CursesDisplayer(scrwin))


if __name__ == '__main__':
    if argv[1:] == ['--version']:
        print version
        exit(0)
    defaults.extend( [
        ( 'parse_dir_interval', 60,
          "how often to rescan the torrent directory, in seconds" ),
        ( 'saveas_style', 2,
          "How to name torrent downloads (1 = rename to torrent name, " +
          "2 = save under name in torrent, 3 = save in directory under torrent name)" ),
        ( 'display_path', 0,
          "whether to display the full path or the torrent contents for each torrent" ),
    ] )
    try:
        configdir = ConfigDir('launchmanycurses')
        defaultsToIgnore = ['responsefile', 'url', 'priority']
        configdir.setDefaults(defaults,defaultsToIgnore)
        configdefaults = configdir.loadConfig()
        defaults.append(('save_options',0,
         "whether to save the current options as the new default configuration " +
         "(only for btlaunchmanycurses.py)"))
        if len(argv) < 2:
            print "Usage: btlaunchmanycurses.py <directory> <global options>\n"
            print "<directory> - directory to look for .torrent files (semi-recursive)"
            print get_usage(defaults, 80, configdefaults)
            exit(1)
        config, args = parseargs(argv[1:], defaults, 1, 1, configdefaults)
        if config['save_options']:
            configdir.saveConfig(config)
        configdir.deleteOldCacheData(config['expire_cache_data'])
        if not os.path.isdir(args[0]):
            raise ValueError("Warning: "+args[0]+" is not a directory")
        config['torrent_dir'] = args[0]
    except ValueError, e:
        print 'error: ' + str(e) + '\nrun with no args for parameter explanations'
        exit(1)

    curses_wrapper(LaunchManyWrapper, config)
    if Exceptions:
        print '\nEXCEPTION:'
        print Exceptions[0]
        print 'please report this to '+report_email
